base-files: ipcalc.sh: Rewrite into shell

pprindeville opened the pull request:
This is a rewrite of ipcalc.sh into shell because there are init.d wrappers that need to be able to invoke some of the internal functions used in ipcalc.sh as library functions.

We also add decimal output of addresses and net masks so that they may easily be used in computation for range checking, etc. in init.d wrappers.

It's a pretty direct rewrite of the original awk code, plus input validation in ip2int(), and some factoring of common terms (we use the complement of the netmask much more often than we use the netmask itself, for instance).

cc: @yogo1212 @Ansuel

pprindeville committed:
base-files: ipcalc.sh: Rewrite in pure shell

Also add better error checking on input.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

pprindeville committed:
base-files: ipcalc.sh: Add support for decimal output

This is useful if you later need to perform numeric range-checking
on addresses, i.e. to see if an address falls inside a CIDR range,
etc. and what interface it corresponds to.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

pprindeville commented:
This is urgently needed functionality to support the ISC-DHCP to Kea migration.

Ansuel commented:
@pprindeville random idea (for the future I'm not asking this) Wonder what would be the size of a C tool that implement these simple thing...

ptpt52 commented:
I like this code

yogo1212 commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -1,100 +1,93 @@
+fi
+
+case "$1" in
+*/*.*)
```
Is this for `192.168.0.12/255.255.255.0`?

I've never seen it written like that and, to my knowledge, the number behind the slash is 'the number of significant bits' (rfc4632) and not a netmask.
Imo, the version on line 35 (two separate arguments) is better suited and this can be dropped.

pprindeville replied:
Yeah, that's a real notation.  Cisco IOS uses it for `ip address x.x.x.x/y.y.y.y` for instance to set an interface's IP address and net mask.

Some people might still be using that and I don't want to break things for them.

yogo1212 replied:
Wow! The people at Cisco really are crazy..

pprindeville replied:
> Wow! The people at Cisco really are crazy..

Guess that would include me.  Spent 5+ years there.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+    echo "$c"
+}
+
+int32() {
```
This needs another name.
Since it's doing `exit 1`, it's basically an assertion.
How about `assert_int32`?

yogo1212 replied:
also, `uint32` - no signs allowed.

pprindeville replied:
Changed to `check_uint32()`.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+# rewrite of /bin/ipcalc.sh from awk to shell for use as a library
+# in init.d wrappers, and better portability (since busybox awk
+# lacks -E support at this time).
+#
```
I'd say, the comment can go :-)

Once this PR is merged, `ipcalc.sh` will be a program using this library (not something this file is a rewrite of).
No need to remind people that this could be implemented using `awk -E`.

pprindeville replied:
Done

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+int32() {
+    local n="$1"
+
+    if [ -n "${n//[0-9]/}" ]; then
```
This is new to me.
Afaik, the canonical way to to check for an integer with POSIX shell is `[ "$n" -eq "$n" ]`. It will be an easier read for those shell aficionados.

pprindeville replied:
It's actually used in a lot of places.  It's idiomatic.  You could use `[0-9a-f]` to check for hex digits, `[0-7]` for octal, etc.  Your suggestion only works with base 10.

yogo1212 replied:
Alright! Interesting to know :-)

For POSIX compat, I would ask for `[:digit:]` - but if busybox has it, whatever.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+    local n="$1"
+
+    if [ -n "${n//[0-9]/}" ]; then
+	echo "Not a decimal integer" >&2
```
Since this is 'buried' and there are no stack traces, would you mind adding `$n` to the output?
With it, it should be easier to find what causes a message (rather than a seemingly random `not a decimal integer`).

pprindeville replied:
Done.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+}
+
+# tedious but portable with busybox's limited shell
+ip2int() {
```
Pew - that's a lot!

How about?

```
        echo "$1" | (
                IFS=. read -r a b c d
                echo "$(( (a << 24) + (b << 16) + (c << 8) + d))"
        )
```

pprindeville replied:
That would accept `10.0.0.1024` and output it as `10.0.4.0`.

yogo1212 replied:
Understood. Still, it's a veeeeery long and repetitive function.

Allow me to make yet another suggestion.

```shell
        ip="$1"

        echo "$ip" | (
                IFS=. read -r a b c d
                [ "$a" -le 256 ] && [ "$b" -le 256 ] && [ "$c" -le 256 ] && [ "$d" -le 256 ] || {
                        echo "malformed IP: $ip" >&2
                        exit 1
                }
                echo "$(( (a << 24) + (b << 16) + (c << 8) + d))"
        )
```

pprindeville replied:
Is the concern simply because it's too long?  I could turn it into a loop, but that might actually slow it down a bit.  Unrolled loops are generally fastest.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+	exit 1
+    fi
+
+    if [ $n -gt 4294967295 ]; then
```
It is clear what is supposed to be done but the constant is difficult to humans to consume.
`$(( 2 ** 32 ))` is much easier to understand (and spot errors in).

pprindeville replied:
C-programmers are familiar with the names in `<limits.h>`.  I'll assign this value to a variable of the same name.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+bitcount() {
+    local c="$1"
+
+    c=$((((c >> 1) & 0x55555555) + (c & 0x55555555)))
```
Need quotes to pass shellcheck.

pprindeville replied:
Problem is that `shellcheck` doesn't know about busybox shell, which is its own thing.

pprindeville replied:
I also don't think that quotes matter inside `$((...))` -- at least not the same way they matter elsewhere when it comes to command-injection, etc.

yogo1212 replied:
You're right about the quotes not being required.
And in fact, no quotes required for arithmetic expansion by shellcheck - I was wrong!

The busybox-compat-issue is very annoying. But most stretches work fine.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,142 @@
+# lacks -E support at this time).
+#
+
+decimal=0
```
This thing bugs me.
It has global variable vibes but has an ambiguous name and is lowercase.
The only thing it does is 'neutering' `int2ip` - it's not used anywhere else.

I recommend against making it part of the shell library.
In `ipcalc.sh`, the variable can still be used to forgo int2ip to achieve the same result.

pprindeville replied:
Okay, makes it less legible but...

yogo1212 commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -1,100 +1,93 @@
-	c=and(rshift(c,16),0x0000ffff)+and(c,0x0000ffff)
-	return c
-}
+#!/bin/sh
```
Always adding quotes is a good idea because checking if there are quotes is a lot easier (and quicker) to do than checking if the code can do without.
I've seen other code in Openwrt being made shellcheck-conformant - assumedly, this should be as well.

pprindeville replied:
Done.

systemcrash commented:
I second some of the comments above, and would like to see a small peppering of code comments as to what various parts are supposed to do, to help us when we look at this in a decade trying to figure something out.

pprindeville committed:
base-files: ipcalc.sh: Add some commentary, etc.

Explain some of the more obscure logic, or where we deviate from
what the original awk code did.  Also, give a count of the usable
addresses on the subnet.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

yogo1212 commented:
> I second some of the comments above, and would like to see a small peppering of code comments as to what various parts are supposed to do, to help us when we look at this in a decade trying to figure something out.

Hi :-)

Well if you see anything you'd like a comment on, point at it.
But I'd still insist that the comment on the role of ipcalc.sh prior to this change should at least be improved.

vgaetera commented:
This breaks backward compatibility:
```bash
# old behavior
> ipcalc.sh 192.168.1.1 24
IP=192.168.1.1
NETMASK=255.255.255.0
NETWORK=192.168.1.0
BROADCAST=192.168.1.255
PREFIX=24

# new behavior
> ipcalc.sh 192.168.1.1 24
Not a dotted quad (24)
IP=192.168.1.1
NETMASK=0.0.0.0
BROADCAST=255.255.255.255
NETWORK=0.0.0.0
PREFIX=0
COUNT=4294967294
```

As a result, this breaks dnsmasq CIDR support:
https://github.com/openwrt/openwrt/pull/13780#issuecomment-1790687076

Possible solutions:
* Restore backward compatibility.
* Modify the dnsmasq init script to fit the new behavior.

yogo1212 commented:
@vgaetera Thanks for debugging the issue and bringing it up!

Imho, the issue lies with the lax handling of the concepts 'netmask' and 'prefix length' in Openwrt.

`netifd` reports this:

```
	"ipv4-address": [
		{
			"address": "192.168.0.157",
			"mask": 24
		}
	],
```

That is a prefix length, not a mask.

`network_get_subnet` joins both on a `/`, turning it into something more reasonable.

`dnsmasq.init`, in turn, uses it like so:

```shell
  network_get_subnet subnet "$net" || return 0
..

  # Override interface netmask with dhcp config if applicable
  config_get netmask "$cfg" netmask "${subnet##*/}"
```

It splits the (previously joined) string back up and, again, a variable called 'netmask' carries a prefix length.
This kind of schizophrenia is exactly what enables these regressions.

eda27e8382f3e5253defab5419e95d4b51f72de3 contains this related change:

```diff
-       config_get netmask "$cfg" netmask
-       [ -n "$netmask" ] || config_get netmask "$net" netmask
+       # Override interface netmask with dhcp config if applicable
+       config_get netmask "$cfg" netmask "${subnet##*/}"
```

That shouldn't have worked.

We can fix it for dnsmasq but who knows who has pieced together their scripts using part of dnsmasq.init.
I'm strongly convinced that integer netmasks should be deprecated and dropped.

A fix for dnsmasq could work like this:

```diff
diff --git a/package/network/services/dnsmasq/files/dnsmasq.init b/package/network/services/dnsmasq/files/dnsmasq.init
index af5ca5822e..74734aeedc 100755
--- a/package/network/services/dnsmasq/files/dnsmasq.init
+++ b/package/network/services/dnsmasq/files/dnsmasq.init
@@ -542,7 +542,7 @@ dhcp_add() {
        [ static = "$proto" ] || return 0
 
        # Override interface netmask with dhcp config if applicable
-       config_get netmask "$cfg" netmask "${subnet##*/}"
+       config_get netmask "$cfg" netmask "$(int2ip "$(prefix_to_netmask "${subnet##*/}")")"
 
        #check for an already active dhcp server on the interface, unless 'force' is set
        config_get_bool force "$cfg" force 0
```

Assuming `prefix_to_netmask` (or similarly named) to be added:

```shell
prefix_to_netmask() {
  echo $((~((1 << (32 - $1)) - 1) & 0xffffffff))
}
```

(also useful for ipcalc.sh)

pprindeville commented:
> This breaks backward compatibility:
> 
> ```shell
> # old behavior
> > ipcalc.sh 192.168.1.1 24
> IP=192.168.1.1
> NETMASK=255.255.255.0
> NETWORK=192.168.1.0
> BROADCAST=192.168.1.255
> PREFIX=24
> 
> # new behavior
> > ipcalc.sh 192.168.1.1 24
> Not a dotted quad (24)
> IP=192.168.1.1
> NETMASK=0.0.0.0
> BROADCAST=255.255.255.255
> NETWORK=0.0.0.0
> PREFIX=0
> COUNT=4294967294
> ```
> 
> As a result, this breaks dnsmasq CIDR support: [#13780 (comment)](https://github.com/openwrt/openwrt/pull/13780#issuecomment-1790687076)
> 
> Possible solutions:
> 
> * Restore backward compatibility.
> * Modify the dnsmasq init script to fit the new behavior.

Where is `ipcalc.sh` being called like this?  And why can't it format the arguments as `192.168.1.1/24` instead, which is standard notation?

pprindeville commented:
> @vgaetera Thanks for debugging the issue and bringing it up!
> 
> Imho, the issue lies with the lax handling of the concepts 'netmask' and 'prefix length' in Openwrt.
> 
> `netifd` reports this:
> 
> ```
> 	"ipv4-address": [
> 		{
> 			"address": "192.168.0.157",
> 			"mask": 24
> 		}
> 	],
> ```
> 
> That is a prefix length, not a mask.

Yikes.

> `network_get_subnet` joins both on a `/`, turning it into something more reasonable.
> 
> `dnsmasq.init`, in turn, uses it like so:
> 
> ```shell
>   network_get_subnet subnet "$net" || return 0
> ..
> 
>   # Override interface netmask with dhcp config if applicable
>   config_get netmask "$cfg" netmask "${subnet##*/}"
> ```
> 
> It splits the (previously joined) string back up and, again, a variable called 'netmask' carries a prefix length. This kind of schizophrenia is exactly what enables these regressions.

That's just plain wrong.  We support:

```
config interface 'lan'
	option ifname 'eth1'
	option proto 'static'
	option ipaddr '192.168.5.1'
	option netmask '255.255.255.0'
```

and that's legit, and the `netmask` really is a netmask.  

I can see wanting to do a pool netmask that's a subset of the interface netmask.

I.e. interface as `192.168.1.1/255.255.255.0` but the pool being `192.168.1.64/255.255.255.192` (i.e. reserving the 2nd fourth of the network for dynamic leases.).

But what they're doing in that line is just plain wrong.  That's what needs to be fixed.

> [eda27e8](https://github.com/openwrt/openwrt/commit/eda27e8382f3e5253defab5419e95d4b51f72de3) contains this related change:
> 
> ```diff
> -       config_get netmask "$cfg" netmask
> -       [ -n "$netmask" ] || config_get netmask "$net" netmask
> +       # Override interface netmask with dhcp config if applicable
> +       config_get netmask "$cfg" netmask "${subnet##*/}"
> ```
> 
> That shouldn't have worked.

Agreed.  The commit is really old.  Pre-git.

> We can fix it for dnsmasq but who knows who has pieced together their scripts using part of dnsmasq.init. I'm strongly convinced that integer netmasks should be deprecated and dropped.

Sorry, what's an "integer netmask"?  Is that something other than a prefix?  Are you talking about dotted-quad netmasks?

> A fix for dnsmasq could work like this:
> 
> ```diff
> diff --git a/package/network/services/dnsmasq/files/dnsmasq.init b/package/network/services/dnsmasq/files/dnsmasq.init
> index af5ca5822e..74734aeedc 100755
> --- a/package/network/services/dnsmasq/files/dnsmasq.init
> +++ b/package/network/services/dnsmasq/files/dnsmasq.init
> @@ -542,7 +542,7 @@ dhcp_add() {
>         [ static = "$proto" ] || return 0
>  
>         # Override interface netmask with dhcp config if applicable
> -       config_get netmask "$cfg" netmask "${subnet##*/}"
> +       config_get netmask "$cfg" netmask "$(int2ip "$(prefix_to_netmask "${subnet##*/}")")"
>  
>         #check for an already active dhcp server on the interface, unless 'force' is set
>         config_get_bool force "$cfg" force 0
> ```
> 
> Assuming `prefix_to_netmask` (or similarly named) to be added:
> 
> ```shell
> prefix_to_netmask() {
>   echo $((~((1 << (32 - $1)) - 1) & 0xffffffff))
> }
> ```
> 
> (also useful for ipcalc.sh)

That should probably go into `ipv4.sh` if at all.  And use `$uint_max` instead of `0xffffffff`.

pprindeville committed:
base-files: ipcalc.sh: Add prefix-to-netmask conversion

Seems like it might be used in other places, so factor it into the
library.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

pprindeville commented:
Just added `prefix_to_netmask()` function and rewrote `ipcalc.sh` to use it.

vgaetera commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,146 @@
+}
+
+# convert prefix into an integer bitmask
+prefix_to_netmask() {
```
You might want to consolidate function naming:
* `int2ip`
* `prefix_to_netmask` > `prefix2netmask`

pprindeville replied:
Done.

vgaetera commented:
> Why not just do [...] and call it good?

Great, but maybe we should include it in this PR since it changes the `ipcalc.sh` behavior.
Otherwise, it will break existing setups that have only address in CIDR notation and no netmask.

vgaetera commented:
> Because `network_get_subnet` returns `ipaddr/prefix` notation which is what `ipcalc.sh` wants.

This makes problematic to override the default netmask:
https://github.com/openwrt/openwrt/blob/ae500e62e2938e112ae1fc6aa7389e8c7b784b13/package/network/services/dnsmasq/files/dnsmasq.init#L542-L543

Note that we need to open a new PR optimizing the dnsmasq init before merging this PR, or include the relevant commit in this PR since it changes the `ipcalc.sh` behavior, otherwise it will break existing setups that have only address in CIDR notation and no netmask.

yogo1212 commented:
> > `network_get_subnet` joins both on a `/`, turning it into something more reasonable.
> > `dnsmasq.init`, in turn, uses it like so:
> > ```shell
> >   network_get_subnet subnet "$net" || return 0
> > ..
> > 
> >   # Override interface netmask with dhcp config if applicable
> >   config_get netmask "$cfg" netmask "${subnet##*/}"
> > ```
> > 
> > It splits the (previously joined) string back up and, again, a variable called 'netmask' carries a prefix length. This kind of schizophrenia is exactly what enables these regressions.
> 
> That's just plain wrong.

I probably wasn't clear about what I meant specifically:
First, the interface information is queried from netifd and put into this format:
`192.168.1.1/24` (ip/prefix-length).
This includes the configuration and/or any DHCP information.

Then, there is: `config_get netmask "$cfg" netmask "${subnet##*/}"`.
According to the comment above it, it was meant to use information including the current DHCP status.
This is puzzling to me. The information from netifd should already include DHCP - if DHCP was ready then.
Also, why would I wait for my router to be assigned an address via DHCP on an interface that I want it to provide a DHCP server itself?
Is there any reason to read the configured netmask if we already have pristine data from netifd? Maybe the line can go altogether.

> We support:
> 
> ```
> config interface 'lan'
> 	option ifname 'eth1'
> 	option proto 'static'
> 	option ipaddr '192.168.5.1'
> 	option netmask '255.255.255.0'
> ```
> 
> and that's legit, and the `netmask` really is a netmask.

What you describe still works because the first query to netifd already includes the netmask configured in UCI (in form of a prefix length).

> I can see wanting to do a pool netmask that's a subset of the interface netmask.
> 
> I.e. interface as `192.168.1.1/255.255.255.0` but the pool being `192.168.1.64/255.255.255.192` (i.e. reserving the 2nd fourth of the network for dynamic leases.).
> 
> But what they're doing in that line is just plain wrong. That's what needs to be fixed.

Agreed.

> > We can fix it for dnsmasq but who knows who has pieced together their scripts using part of dnsmasq.init. I'm strongly convinced that integer netmasks should be deprecated and dropped.
> 
> Sorry, what's an "integer netmask"? Is that something other than a prefix? Are you talking about dotted-quad netmasks?

Sorry, unclear. I meant the former (prefix length).

> > Assuming `prefix_to_netmask` (or similarly named) to be added:
> > ```shell
> > prefix_to_netmask() {
> >   echo $((~((1 << (32 - $1)) - 1) & 0xffffffff))
> > }
> > ```
>>
> > (also useful for ipcalc.sh)
> 
> That should probably go into `ipv4.sh` if at all. And use `$uint_max` instead of `0xffffffff`.
> 
> Or you could do:
> 
> ```
> prefix_to_netmask() {
>     echo "$(((~($uint_max >> $1)) & $uint_max))"
> }
> ```

I just copied the calculation from this PR. You're right. Using `$uint_max` is nicer.
Regarding `(~($uint_max >> $1))` vs.  `~((1 << (32 - $1)) - 1)`, I have no strong preference. Your variant is easier to read.

yogo1212 commented:
> This is puzzling to me. The information from netifd should already include DHCP - if DHCP was ready then.
Also, why would I wait for my router to be assigned an address via DHCP on an interface that I want it to provide a DHCP server itself?

Regarding `config_get netmask "$cfg" netmask`:
`$cfg` is the DHCP pool in UCI. `netmask` is not documented in the wiki to be an option.
Even if it was supposed to work at some time in the past, it can be in direct conflict with the other information (from netifd).
Now, it's not used in a way that would make it effective anywhere.

It should probably be addressed in @vgaetera 's latest PR on the matter:
https://github.com/openwrt/openwrt/pull/13870

pprindeville commented:
>> [...]
> > That's just plain wrong.
> 
> I probably wasn't clear about what I meant specifically: First, the interface information is queried from netifd and put into this format: `192.168.1.1/24` (ip/prefix-length). This includes the configuration and/or any DHCP information.

Yes.  It doesn't support non-contiguous netmasks (i.e. where there are zeros anywhere but in the LSBs), but I don't know if anyone uses that still (footnote).

> Then, there is: `config_get netmask "$cfg" netmask "${subnet##*/}"`. According to the comment above it, it was meant to use information including the current DHCP status. This is puzzling to me. The information from netifd should already include DHCP - if DHCP was ready then. Also, why would I wait for my router to be assigned an address via DHCP on an interface that I want it to provide a DHCP server itself? Is there any reason to read the configured netmask if we already have pristine data from netifd? Maybe the line can go altogether.

That's not what it's trying to do.  It's trying to have a longer prefix than the interface's subnet, i.e. use the lowest 1/2, 1/4, 1/8th, etc. depending on how many more bits longer than the interface prefix is.

I understand the utility of doing this... What I don't understand is arbitrarily restricting it to the very top of the subnet.  Maybe @jow- can chime in as to what the original intention was, and if it's worth carrying forward?

I just pushed a commit to propagate that, but with @jow-'s blessing I'd rather just drop support for `netmask` within DHCP parameters.

As another footnote: you can use `start` and `limit` to place the pool anywhere inside of the interface's subnet, so it's of little consequence other than as a convenience.  If you do use `netmask` (which is actually a *prefix*, so this is confusing) and a non-zero `start`, then you've increased your probability of shooting yourself in the foot... I'd argue that if you have a `netmask` set, then you should *ignore* `start` and `limit` and just consume the entire fractional subnet.

If we detect this parameter and it's anything other than what the interface's prefix is set to, we could burp a warning that this is deprecated and unsupported...

> [snip]

> > > We can fix it for dnsmasq but who knows who has pieced together their scripts using part of dnsmasq.init. I'm strongly convinced that integer netmasks should be deprecated and dropped.
> > 
> > 
> > Sorry, what's an "integer netmask"? Is that something other than a prefix? Are you talking about dotted-quad netmasks?
> 
> Sorry, unclear. I meant the former (prefix length).

Okay, yeah.  That's a problem.  People copying code that introduced dubious notation...  While @dangowrt and I were rewriting `isc-dhcp` to support the notation used by `dnsmasq`, we found some things that just didn't make sense and needed to go for the sake of KISS (and I'd add -C for "consistent", too).

> [snip]
> 
> I just copied the calculation from this PR. You're right. Using `$uint_max` is nicer. Regarding `(~($uint_max >> $1))` vs. `~((1 << (32 - $1)) - 1)`, I have no strong preference. Your variant is easier to read.

See the 2nd to last commit.

pprindeville commented:
Or we could just merge PR #13879 today and rebase this PR (yes, there will be conflicts but not a lot).

jow- commented:
> Regarding config_get netmask "$cfg" netmask:
$cfg is the DHCP pool in UCI. netmask is not documented in the wiki to be an option.
Even if it was supposed to work at some time in the past, it can be in direct conflict with the other information (from netifd).
Now, it's not used in a way that would make it effective anywhere.

This option allows overriding the netmask used for DHCP with one deviating from the prefix used by the interface the pool is bound to. For example to use a range of `192.168.1.100-192.168.1.200` in subnet `192.168.1.0/24` while the related interface uses a larger subnet such as `192.168.0.0/22`.

pprindeville commented:
> > Regarding config_get netmask "$cfg" netmask:
> > $cfg is the DHCP pool in UCI. netmask is not documented in the wiki to be an option.
> > Even if it was supposed to work at some time in the past, it can be in direct conflict with the other information (from netifd).
> > Now, it's not used in a way that would make it effective anywhere.
> 
> This option allows overriding the netmask used for DHCP with one deviating from the prefix used by the interface the pool is bound to. For example to use a range of `192.168.1.100-192.168.1.200` in subnet `192.168.1.0/24` while the related interface uses a larger subnet such as `192.168.0.0/22`.

Okay, but what value of option 1 does the client get?  `255.255.255.0` or `255.255.252.0`?

yogo1212 commented:
To fix the bug, we could start with something like this:

```diff
diff --git a/package/network/services/dnsmasq/files/dnsmasq.init b/package/network/services/dnsmasq/files/dnsmasq.init
index d356a21f81..e592c2efb4 100755
--- a/package/network/services/dnsmasq/files/dnsmasq.init
+++ b/package/network/services/dnsmasq/files/dnsmasq.init
@@ -583,7 +583,7 @@ dhcp_add() {
        nettag="${networkid:+set:${networkid},}"
 
        # make sure the DHCP range is not empty
-       if [ "$dhcpv4" != "disabled" ] && ipcalc "${subnet%%/*}" "$netmask" "$start" "$limit" ; then
+       if [ "$dhcpv4" != "disabled" ] && ipcalc "${subnet%%/*}/$netmask" "$start" "$limit" ; then
                [ "$dynamicdhcpv4" = "0" ] && END="static"
 
                xappend "--dhcp-range=$tags$nettag$START,$END,$NETMASK,$leasetime${options:+ $options}"
```

It still has a prefix length in a variable called 'netmask' most of the time but at least the bug is fixed (`192.168.0.1/24` and `
192.168.0.1/255.255.255.0` are both accepted by ipcalc).

pprindeville commented:
Requires PR #13765.

pprindeville commented:
> To fix the bug, we could start with something simple like this:
> 
> ```diff
> diff --git a/package/network/services/dnsmasq/files/dnsmasq.init b/package/network/services/dnsmasq/files/dnsmasq.init
> index d356a21f81..e592c2efb4 100755
> --- a/package/network/services/dnsmasq/files/dnsmasq.init
> +++ b/package/network/services/dnsmasq/files/dnsmasq.init
> @@ -583,7 +583,7 @@ dhcp_add() {
>         nettag="${networkid:+set:${networkid},}"
>  
>         # make sure the DHCP range is not empty
> -       if [ "$dhcpv4" != "disabled" ] && ipcalc "${subnet%%/*}" "$netmask" "$start" "$limit" ; then
> +       if [ "$dhcpv4" != "disabled" ] && ipcalc "${subnet%%/*}/$netmask" "$start" "$limit" ; then
>                 [ "$dynamicdhcpv4" = "0" ] && END="static"
>  
>                 xappend "--dhcp-range=$tags$nettag$START,$END,$NETMASK,$leasetime${options:+ $options}"
> ```
> 
> It still has a prefix length in a variable called 'netmask' most of the time but at least `192.168.0.1/24` and `192.168.0.1/255.255.255.0` are both accepted by ipcalc).

That's basically all PR #13765 does.

yogo1212 commented:
> That's basically all PR https://github.com/openwrt/openwrt/pull/13765 does.

Do you mean PR #13765 ? (this one)
Perhaps you could just include the change here and put the commit right in front of the rewrite.

pprindeville commented:
> > Regarding config_get netmask "$cfg" netmask:
> > $cfg is the DHCP pool in UCI. netmask is not documented in the wiki to be an option.
> > Even if it was supposed to work at some time in the past, it can be in direct conflict with the other information (from netifd).
> > Now, it's not used in a way that would make it effective anywhere.
> 
> This option allows overriding the netmask used for DHCP with one deviating from the prefix used by the interface the pool is bound to. For example to use a range of `192.168.1.100-192.168.1.200` in subnet `192.168.1.0/24` while the related interface uses a larger subnet such as `192.168.0.0/22`.

Just to be clear (since it's not documented), is the `netmask` parameter in DHCP's config a dotted-quad or an integer between 0..32?

pprindeville commented:
> Perhaps you could just include the change here and put the commit right in front of the rewrite.

It needs to be fixed regardless of when the `ipcalc.sh` rewrite goes in.  I'd rather not create an unnecessary dependency.

pprindeville commented:
> Ahh - got it! #13885 Sorry for the spam

It was PR #13879 but now it's #13885.

yogo1212 reviewed (approved):
I approve of these changes the way they are now and I look forward to seeing this PR merged.

Let me promote, once more, the idea of making `ip2int` a little less bulky.
Maybe there is a way we could use a function to encapsulate the select statement?

An implementation like this might is much easier to consume, check, and understand:
```
echo "$ip" | (
  IFS=. read -r a b c d
  # maybe check for negative numbers as well
  if [ "$a" -le 256 ] && [ "$b" -le 256 ] && [ "$c" -le 256 ] && [ "$d" -le 256 ]; then
    echo "malformed IP: $ip" >&2
    return 1
  fi
  echo "$(( (a << 24) + (b << 16) + (c << 8) + d))"
)
```

pprindeville commented:
> I approve of these changes the way they are now and I look forward to seeing this PR merged.
> 
> Let me promote, once more, the idea of making `ip2int` a little less bulky. Maybe there is a way we could use a function to encapsulate the select statement?
> 
> An implementation like this might is much easier to consume, check, and understand:
> 
> ```
> echo "$ip" | (
>   IFS=. read -r a b c d
>   # maybe check for negative numbers as well
>   if [ "$a" -le 256 ] && [ "$b" -le 256 ] && [ "$c" -le 256 ] && [ "$d" -le 256 ]; then
>     echo "malformed IP: $ip" >&2
>     return 1
>   fi
>   echo "$(( (a << 24) + (b << 16) + (c << 8) + d))"
> )
> ```

For a start, it would be `-le 255` ...

systemcrash commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,146 @@
+
+# tedious but portable with busybox's limited shell
+# we check each octet to be in the range of 0..255,
+# and also make sure there's no extaneous characters
```
`# and also make sure there are no extraneous characters`

pprindeville replied:
Changed

systemcrash commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,146 @@
+uint_max=4294967295
```
Is there an advantage to include the shebang here? I know this is an included/sourced script, but would it ever be executed directly?

pprindeville replied:
It only includes function definitions, so nothing that would get executed because there’s no main body.

pprindeville committed:
base-files: ipcalc.sh: Add netmask2prefix function

Also add is_contiguous to check if it's a valid netmask.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

pprindeville commented:
Can everyone have another look?  In the places where a function was being called in a subshell i.e. `$(...)` I've added an `exit 1` if that returned non-zero.

I also added a `netmask2prefix()` function and `is_contiguous()` to enforce sane netmasks.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,166 @@
+    local ip="$1" n ret=0
+
+    case "$ip" in
+    [0-9].*)
```
This correctly parses address bytes in the octal notation but not hexadecimal - both have been supported by the version of ipcalc.sh prior to this (and are supported by tools like ping from iputils).

pprindeville replied:
It actually doesn't parse octal.  And no one has used octal dotted-quad notation in 40 years.

yogo1212 replied:
My bad, I overlooked the [1-9] at the start.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,166 @@
+uint_max=4294967295
+
+# Common error handler
+__die() {
```
I've never really been happy with the `exit` inside shell library code.
In a work project, I would generally not allow it. Symbols from a shell library should also have a common prefix, etc. With free software mostly being volunteer work, I try not to nag too much.

But I'm bringing it up again because it hinders error handling.
It's not possible to catch an error unless the function is called in a subshell (which you don't suspect you'd have to do).

With ipcalc.sh in-between and most functions at least going through `$()` (which isn't run in a new process but handles the `exit`) it isn't hugely dramatic **now** but if anyone either merges the range the calculation into `ipv4.sh` (with callers no longer going through ipcalc.sh) or decides to use any other library functions directly, their processes just die.

I started "fumbling with the exceptionally stable ipcalc.sh" ('stable' for not having been maintained in x years) because a mistake in the configuration of an (otherwise unrelated) interface (like `lan`) can bring down local name resolution on the router itself (and, thus, an important part of the internet connection).
In fact, a configuration or runtime error on a separate interface (e.g. `guest`) could bring down `br-lan` and the name resolution.

Either it has to be absolutely clear that `check_uint32`, `ip2int`, `netmask2prefix`, and `netmask2prefix` can cause the process to exit (which is why I proposed the `assert_` prefix) or these functions have to fail with a return value (my preferred option because otherwise, there'd have to non-`assert_` alternatives for them sooner or later).
`ipv4.sh` looks very juicy and I suspect other packages want to use it as well - we can't assume here that they're all single-purpose and that calling exit has no side effects.

pprindeville replied:
> It's not possible to catch an error unless the function is called in a subshell (which you don't suspect you'd have to do).

Actually, if I'm calling `func args` and ignoring the return value, then that would run in the shell.

But if I'm calling `$(func args)` and capturing the output, then that runs in a sub shell.  All of the functions in `ipv4.sh` *except* `is_contiguous()` are expected to run in a sub shell.  That's why `is_contiguous()` returns true/false but never exits.

> With `ipcalc.sh` in-between and most functions at least going through `$()` (which isn't run in a new process [...])

Again, all but one of the functions will get called in a sub shell.

>  [...] or decides to use any other library functions directly, their processes just die.

Again, that depends on whether they're using `eval` or `$(...)` to call those functions.

I could have functions that return their results by polluting the callers environment (and hence never have to run in a sub shell, because they generate no output) but you didn't seem to like that idea either.

>` ipv4.sh` looks very juicy and I suspect other packages want to use it as well - we can't assume here that they're all single-purpose and that calling exit has no side effects.

Indeed, I wrote it because I'll need to use it to do the rewrite of `dhcpd.init` in `isc-dhcp` to output Kea configs instead.

yogo1212 replied:
> Actually, if I'm calling `func args` [..] then that would run in the shell.

I'm well aware.
It's the

> and ignoring the return value

that I'm worried about.

Here's an example that doesn't ignore the return value:

```shell
f() { exit 1; }
f || echo unreachable
```

Here's a (imo) reasonable way to use `check_uint32`:

```shell
check_uint32 "-1" || {
  # imposible to reach
}
```

> I could have functions that return their results by polluting the callers environment (and hence never have to run in a sub shell, because they generate no output) but you didn't seem to like that idea either.

I'm not sure what you mean. Yes, I don't like library functions terminating my programs unexpectedly AND I don't like them unexpectedly overwriting my variables or functions too.
I'm convinced that is neither unreasonable to want nor impossible to achieve.

> All of the functions in ipv4.sh except is_contiguous() are expected to run in a sub shell.

That fact would at least have to be  be documented somewhere but notes like that are easily overlooked.
If that is true, you can use `f() ( .. )` (rather than `f() { .. }`) to define those functions and the problem goes away.
People working on ipv4.sh in the future should be made aware as well, which I hope a comment on `__die` suffices for.

```shell
f() ( exit 1; )
f || echo reachable
```

pprindeville replied:
> > Actually, if I'm calling `func args` [..] then that would run in the shell.
> 
> I'm well aware. It's the
> 
> > and ignoring the return value
> 
> that I'm worried about.
> 
> Here's an example that doesn't ignore the return value:
> 
> ```shell
> f() { exit 1; }
> f || echo unreachable
> ```

This is a red herring.

The only functions that `exit 1` are the ones that generate results not by changing external state, but merely by outputting their calculation on `stdout`.  They will *always* be called via `$(...)` and hence as a sub shell.

The example you cite above just isn't how they will be invoked.

> Here's a (imo) reasonable way to use `check_uint32`:
> 
> ```shell
> check_uint32 "-1" || {
>   # imposible to reach
> }
> ```

Per your suggestion, I changed this to (1) `assert_uint32` and (2) it does a `return 1` not an `exit 1`.  That change was already present when you made this comment.

>  > I could have functions that return their results by polluting the callers environment (and hence never have to run in a sub shell, because they generate no output) but you didn't seem to like that idea either.
> 
> I'm not sure what you mean. Yes, I don't like library functions terminating my programs unexpectedly AND I don't like them unexpectedly overwriting my variables or functions too. I'm convinced that is neither unreasonable to want nor impossible to achieve.

The problem is that you have a scarcity of means here.  *If* we're putting the result of a calculation onto `stdout`, then by necessity we're running the function in a sub shell, *and* we need a means to generate an error that doesn't overload `stdout`.  This is shell, a 40 year old language.  We're not going to find a previously overlooked methodology for handling this usage case.

The mechanism to communicate errors in sub shells that doesn't compromise `stdout` *is* the `exit` status.

> > All of the functions in ipv4.sh except is_contiguous() are expected to run in a sub shell.
> 
> That fact would at least have to be be documented somewhere but notes like that are easily overlooked. If that is true, you can use `f() ( .. )` (rather than `f() { .. }`) to define those functions and the problem goes away. People working on ipv4.sh in the future should be made aware as well, which I hope a comment on `__die` suffices for.
> 
> ```shell
> f() ( exit 1; )
> f || echo reachable
> ```

yogo1212 replied:
> The mechanism to communicate errors in sub shells that doesn't compromise stdout is the exit status.

'in sub shells". If there is no subshell, the argument is moot. I'm trying to tell you that it's hard to control for that reliably for users now and future code with the current code.

> Per your suggestion, I changed this to (1) assert_uint32 and (2) it does a return 1 not an exit 1.

Yes, I suggested both *independently* and gave possible reasons for doing either. You forgo the reasoning, do both and tell me that I should be happy now in a "I did what you said even though I think it's stupid. Aren't you happy now?" kind of way.

I don't know about you but I'm done talking about bike-sheds and red herrings.
Please do the rest of this without me.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,166 @@
+check_uint32() {
+    local n="$1"
+
+    [ -z "$n" -o -n "${n//[0-9]/}" ] && __die "Not a decimal integer ($n)"
```
For prefixes, awk's `int` was used which only accepts decimals - and it's reasonable to carry over.
But for doing calculations, it is reasonable to write integer addresses using the hexadecimal notation.
This would apply to `int2ip`, `is_contiguous`, and `netmask2prefix`.

yogo1212 replied:
It is easy enough to use shell perform the conversion `$((ip))` while invoking the function.
Still, this would bug me less if it were named to describe more clearly what it's doing.
`check_decimal_uint32` would be an appropriate, imo.

I could also see two versions existing: `require_decimal_uint32` and `convert_to_decimal_uint32` (names to point out difference).

Linux tools tend to be fairly relaxed with the other numeral systems: `ip a add 10.0xa.0.012/0x18 ..`.
No idea, though, how the landscape looks like in Openwrt.

pprindeville replied:
> For prefixes, awk's `int` was used which only accepts decimals - and it's reasonable to carry over. But for doing calculations, it is reasonable to write integer addresses using the hexadecimal notation. This would apply to `int2ip`, `is_contiguous`, and `netmask2prefix`.

If I'm doing a calculation, then I can use `$((0xffff0000))` to do the base conversion for me.

pprindeville replied:
> It is easy enough to use shell perform the conversion `$((ip))` while invoking the function. Still, this would bug me less if it were named to describe more clearly what it's doing. `check_decimal_uint32` would be an appropriate, imo.

We're getting into bike-shedding territory.

If the names become too long, then the calling code will start to look ungainly.

I say we put it out there and see what the reactions are.

pprindeville replied:
I've renamed it `assert_uint32` and it returns true/false, and if it fails, one should `exit 1`.

yogo1212 replied:
Look, we both want this merged and I don't want to contribute to any misery. You don't have to change anything to please me. You are much more 'Openwrt' than I am.
You asked me for a review and I'm offering some time of my life for feedback.

Assertions are generally expected to end the current control flow, which is why I proposed to include the concept in the name - people should at least get a chance to learn that their program can halt.
Changing the name to `assert_` and going from exit to return values at the same time is nonsensical from that perspective. Now, people can within a reasonable mind expect that their control flow will be halted for illegal values when, in reality, it won't.

I've also pointed out that passing the argument to `check_uint32` through shell `$((ip))`. The point is that, unless your look how `check_uint32` is implemented, you don't know you need to be doing that.
The same goes for the interrupted control flow. That is not bike-shedding - it's API design 101.

`check_uint32` communicates 'I can call this to check if my integer is in range'. It doesn't communicate numeral system requirements, it doesn't communicate that the ended control flow, it doesn't communicate the usage of stdout.

What you would intuitively want to with `check_uint32` do is something like this:

```shell
if ! check_uint32 "$inside_tunnel_netmask"; then
  echo "server-supplied netmask is invalid: $inside_tunnel_netmask" >&2
  continue
fi
```

In it's current form, I would only be getting `Not normalized notation` and have to go looking where the number came from.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,166 @@
+    local n
+    n="$(check_uint32 "$1")" || exit 1
+
+    echo "$((n >> 24)).$(((n >> 16) & 255)).$(((n >> 8) & 255)).$((n & 255))"
```
"ip" is a quad - "netmask" is an integer.
Not dramatic - but easy to improve now.
How about `ip_from_str`/`ip_to_str`?
`ip2str`/`str2ip`?

pprindeville replied:
`ip2int` was what was in the original awk code and no one had a problem with it, but I will rename it now so we can move forward.

yogo1212 replied:
What kind of reasoning is that? It's an incremental process.
If no one had a problem with the old code, why this PR?
Anyway, I'm out.

pprindeville commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -1,100 +1,122 @@
+case "$1" in
+*/*.*)
+    # data is n.n.n.n/m.m.m.m format, like on a Cisco router
+    ipaddr=$(ip2int "${1%/*}") || exit 1
```
Added.

yogo1212 commented on package/base-files/files/lib/functions/ipv4.sh:
```diff
@@ -0,0 +1,185 @@
+    fi
+
+    if [ "$((n + 0))" != "$n" ]; then
+	echo "Not normalized notation (%s)" "$n" >& 2
```
echo/printf mixup + typo at the end of the line (space between `& 2`).

pprindeville replied:
The space doesn't make a difference but I changed it anyway.

pprindeville commented:
@systemcrash Have your comments been addressed?

systemcrash commented:
My concerns here are least important. We can always add fixes later. I'm
happy if you're happy.

dave14305 commented:
Since https://github.com/openwrt/openwrt/commit/af64898c2617babcf27ba2ae8c07dc801c208964 went in recently, some dhcp scenarios involving a netmask are broken in main, waiting for this PR which will accept ip/netmask format.

https://forum.openwrt.org/t/dhcp-broken-with-snapshot/177849/

pprindeville commented:
Complete rewrite per @jow-'s suggestion on a different PR, to always return a status code, and to deposit results into a variable (pass-by-reference).

vgaetera commented:
It looks like you need to add conversion to both functions:
```bash
> . /lib/functions/ipv4.sh
> netmask2prefix prefix 255.255.255.0; echo ${prefix}
Not a decimal integer (255.255.255.0 )
> prefix2netmask netmask 24; echo ${netmask}
4294967040
```

vgaetera commented:
Testing:
```bash
> . /lib/functions/ipv4.sh
> netmask2prefix prefix 255.255.255.0; echo ${prefix}
Not a decimal integer (255.255.255.0 )

> prefix2netmask netmask 24; echo ${netmask}
4294967040
```

pprindeville commented:
> Testing:
> 
> ```shell
> > . /lib/functions/ipv4.sh
> > netmask2prefix prefix 255.255.255.0; echo ${prefix}
> Not a decimal integer (255.255.255.0 )
> 
> > prefix2netmask netmask 24; echo ${netmask}
> 4294967040
> ```

Oh, yeah. You’ll need to call `str2ip` on the dotted quad first!

pprindeville committed:
base-files: ipcalc.sh: Add tests for unroutable, etc

See RFC-1918, RFC-3927, and RFC-1122.

Signed-off-by: Philip Prindeville <philipp@redfish-solutions.com>

pprindeville commented:
@nbd168 @hauke @neocturne @kaloz @stintel @karlp @KanjiMonster @dwmw2 @ynezz @neheb @blogic @aparcar @yousong Casting a still wider net... Sorry for the spam but this seems to be stuck in limbo.

pprindeville commented:
Thanks!

yogo1212 commented:
So it was possible to make this safer to use.

Regardless of our personal disagreement, I'm happy you got this merged.
Thanks for putting in the work and pushing this through!
It's a good thing to have.

pprindeville commented:
> So it was possible to make this safer to use.
> 
> Regardless of our personal disagreement, I'm happy you got this merged. Thank you for putting in the work and pushing this through! It's a good thing to have.

I hope it’s broadly useful.
